### PR TITLE
Add ShardingClientBuilder

### PR DESCRIPTION
**Pull request by quanticc**
**Description:** Implements building coordinated sharding clients.
**Justification:** Reduce boilerplate currently required to perform this operation.

This could be the first step in fulfilling #421. We would include ``Mono<GatewayConnection>`` in a separate PR, suggested in #483 

1. Add a builder to make many clients at once, coordinated by default, sharing stores and router.
2. Move some of the resource creation out of DiscordClientBuilder#build to allow external configuration from ShardingClientBuilder, in particular JacksonResourceProvider and ReactorResourceProvider.
3. Simplify the creation of these resources by providing defaults, this means DiscordWebClient sets its default headers inside, which in turn causes VersionUtil to be moved and renamed to GitProperties at common module.
4. Promote coordinated shard + registry from test. Given the feedback from Shadorc it is stable to use as default.
5. Add docs, so much left!
6. Update the tests showing the reduction in boilerplate

### CODE REVIEW

File: common/src/main/java/discord4j/common/JacksonResourceProvider.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+            .setVisibility(PropertyAccessor.FIELD, JsonAutoDetect.Visibility.ANY)
+            .registerModules(new PossibleModule(), new Jdk8Module());
+
+    private final ObjectMapper objectMapper;
+
+    /**
+     * Create a default {@link com.fasterxml.jackson.databind.ObjectMapper} that allows any field visibility,
+     * registers modules to handle Discord4J specific mappings and ignores unknown properties.
+     */
+    public JacksonResourceProvider() {
```
Comment (GrandPanda) [inline common/src/main/java/discord4j/common/JacksonResourceProvider.java]: Can this call the overloaded constructor with the used `mapper`?

Comment (quanticc) [reply common/src/main/java/discord4j/common/JacksonResourceProvider.java]: Sure, looks cleaner.

File: common/src/main/java/discord4j/common/ReactorResourceProvider.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+/**
+ * Provides Reactor Netty resources like an {@link reactor.netty.http.client.HttpClient} that can be customized and
+ * reused across the application.
+ * <p>
+ * Allow a user to externally manage the connection pool through a custom
+ * {@link reactor.netty.resources.ConnectionProvider}, and custom event loop threads using a
+ * {@link reactor.netty.resources.LoopResources}.
+ */
+public class ReactorResourceProvider {
```
Comment (GrandPanda) [inline common/src/main/java/discord4j/common/ReactorResourceProvider.java]: The naming here isn't clear to me. In ClientBuilder it is referred to as the "rest resource provider."

File: common/src/main/java/discord4j/common/GitProperties.java [hunk 3]
```diff
@@ -15,26 +15,38 @@
...
 
 import java.io.IOException;
 import java.io.InputStream;
 import java.util.Properties;
 
-public final class VersionUtil {
+/**
+ * Provide information about the Git repository version captured at build time.
+ */
+public class GitProperties {
```
Comment (danthonywalker) [inline common/src/main/java/discord4j/common/GitProperties.java]: I'm pretty sure we use singular naming throughout the project. If not we should change that. https://stackoverflow.com/a/15756009/5302085

Comment (danthonywalker) [inline common/src/main/java/discord4j/common/GitProperties.java]: Also in general I feel the design of this class can be improved. I feel it should be an enum, `System.out.println(GitPropery.APPLICATION_NAME.toString())` would print what you would expect intuitively, while, for myself, I've done this in the past `System.out.println(GitProperty.APPLICATION_NAME)` and it wasn't obvious why. The reliance of `getProperties` to get the correct value I feel is confusing and we can be better. In fact, we could just make its instantiation static, so it's cached, for the enums to work off of.

File: common/src/main/java/discord4j/common/ReactorResourceProvider.java (hunk 2 above)
Comment (danthonywalker) [reply common/src/main/java/discord4j/common/ReactorResourceProvider.java]: +1 for Http

File: core/src/main/java/discord4j/core/DiscordClientBuilder.java [hunk 4]
```diff
@@ -198,8 +204,8 @@ public DiscordClientBuilder setShardCount(@Nullable Integer shardCount) {
     }
 
     /**
-     * Get the current {@link discord4j.store.api.service.StoreService} factory, used to create
-     * {@link discord4j.store.api.Store} instances on login to cache entities.
+     * Get the current {@link StoreService} factory, used to create {@link discord4j.store.api.Store} instances on
```
Comment (danthonywalker) [inline core/src/main/java/discord4j/core/DiscordClientBuilder.java]: Remove fully qualified import.

File: core/src/main/java/discord4j/core/DiscordClientBuilder.java [hunk 5]
```diff
@@ -211,8 +217,8 @@ public StoreService getStoreService() {
     }
 
     /**
-     * Set a new {@link discord4j.store.api.service.StoreService} to this builder, used to create
-     * {@link discord4j.store.api.Store} instances on login to cache entities.
+     * Set a new {@link StoreService} to this builder, used to create {@link discord4j.store.api.Store} instances on
```
Comment (danthonywalker) [inline core/src/main/java/discord4j/core/DiscordClientBuilder.java]: Remove fully qualified import.

File: core/src/main/java/discord4j/core/DiscordClientBuilder.java [hunk 6]
```diff
@@ -69,11 +67,11 @@
...
 /**
- * Builder suited for creating a {@link discord4j.core.DiscordClient}.
+ * Builder suited for creating a {@link DiscordClient}.
  * <p>
  * Allows creating a Discord client for a single shard with many configurable options. You can reuse this builder for
  * sharding as long as you set the {@link #setShardIndex(Integer)} and {@link #setShardCount(Integer)}
- * options, or {@link #setIdentifyOptions(discord4j.gateway.IdentifyOptions)}.
+ * options, or {@link #setIdentifyOptions(IdentifyOptions)}.
  */
 public final class DiscordClientBuilder {
```
Comment (danthonywalker) [inline core/src/main/java/discord4j/core/DiscordClientBuilder.java]: I feel at this point DiscordClientBuilder is becoming a god object. This particular review is not a hindrance to pushing this PR but more to comment in that we need to seriously reconsider this entire design at a later date.

File: core/src/main/java/discord4j/core/shard/ShardingClientBuilder.java [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+import java.util.function.Predicate;
+
+/**
+ * Builder suited for creating a sequence of {@link discord4j.core.DiscordClient} each dedicated to a shard, while
+ * allowing resource sharing and login coordination.
+ * <p>
+ * Creates multiple Discord clients with many configurable options. It uses a recommended amount of shards but can be
+ * overridden through {@link #setShardCount(Integer)}.
+ */
+public class ShardingClientBuilder {
```
Comment (danthonywalker) [inline core/src/main/java/discord4j/core/shard/ShardingClientBuilder.java]: We should just remove getters on both this builder and DiscordClientBuilder. Getters on builders is an unusual pattern, at least I have not seen an example of it in a popular/enterprise project or even in the JDK itself (refer to HttpClient.Builder)

Plus I don't see much use cases considering you can't multi-thread these anyways.

File: common/src/main/java/discord4j/common/GitProperties.java (hunk 3 above)
Comment (quanticc) [reply common/src/main/java/discord4j/common/GitProperties.java]: I agree, this was just a package move. We could improve on it later.

**Pull request merged**